### ISSUE TITLE
worker: watch closed or error running the worker process

### ISSUE THREAD

**Issue by itssimon** [pachyderm/pachyderm#3894]
**What happened?**:

My pipeline is stuck in merging state after processing all datums. The logs show that uploading output failed for a few datums.

```
[2019-07-04 10:43:33.871016741+10:00] finished running user code after 1.485912319s
[2019-07-04 10:43:33.871221542+10:00] starting to upload output
[2019-07-04 10:43:33.888924599+10:00] errored uploading output after 17.617342ms: rpc error: code = Unknown desc = Put https://pachyderm.blob.core.windows.net/pachyderm//pach/block/xxx?blockid=xxx&comp=block: read tcp 10.95.8.146:36148->20.38.112.36:443: read: connection reset by peer
[2019-07-04 10:43:33.889482737+10:00] failed processing datum: rpc error: code = Unknown desc = Put https://pachyderm.blob.core.windows.net/pachyderm//pach/block/xxx?blockid=xxx&comp=block: read tcp 10.95.8.146:36148->20.38.112.36:443: read: connection reset by peer, retrying in 0s
[2019-07-04 10:43:33.889582872+10:00] starting to download data
[2019-07-04 10:43:33.903355598+10:00] finished downloading data after 13.701977ms
[2019-07-04 10:43:33.903727293+10:00] beginning to run user code
```

It says it's retrying but never logs `finished uploading output after ...` subsequently. Instead it seems to proceed to download data for the next datum straight away.

While stuck in merging state the following log entries are produced:

```
error downloading chunk 589 from worker at 10.95.8.205 (rpc error: code = Unknown desc = key 589 not found in cache), falling back on object storage
...
worker: watch closed or error running the worker process: acquire/process/merge datums for job eba2fc917a1b4c1799058190bd1f7f02 exited with err: rpc error: code = Unknown desc = tagGetter: tag a539d7f532a36641dea3f005539f10a0fbfa8218a3b9fa9981b1921b55091333d66e not found; retrying in 12.468383509s
```

Another pipeline was running simultaneously and got stuck in running state at the same time. The logs show the same type of message as above:

```
worker: watch closed or error running the worker process: acquire/process/merge datums for job d993876f41244aedaf6f31ac8a253417 exited with err: error claiming/processing chunk: tagGetter: tag b46acb33f06040c11c13a60b35bb097fb17f0d71d20bb098c761b668b0c9a479fd55 not found; retrying in 11.303107507s
```

**How to reproduce it (as minimally and precisely as possible)?**:

This seems to happen sporadically and I'm not sure how to reproduce reliably.

**Anything else we need to know?**:

I can't get things back to working order even after restarting `pachd` and updating the pipelines. Something must be corrupted now.

**Environment?**:
- Kubernetes version: 1.14
- Pachyderm CLI version: 1.9.0
- Cloud provider: Azure

**Comment by JoeyZwicker** [comment]
Just to double check @itssimon, your pachd version is 1.9.0 too?

@brycemcanally solved a few bugs of this type before 1.9, so he might be able to shed more light or ask some followup questions.

**Comment by itssimon** [comment]
Good call, I must've been using an older version of the protobufs. Thanks! Closing...

**Comment by itssimon** [comment]
Actually, I was too quick to close this. I have re-deployed with up-to-date pachctl and recreated all repos and pipelines. Still running into this issue where pipelines get stuck in merging state with the above mentioned log entries being produced.

This time it happened after updating two pipelines with --reprocess. Both are stuck in merging.

**Comment by JoeyZwicker** [comment]
@msteffen @brycemcanally Any ideas on this?

@itssimon Can you give us fuller logs or are they all just this same message?

**Comment by itssimon** [comment]
The log messages above are all there is. They just keep repeating every 20 seconds.

**Comment by itssimon** [comment]
I'm hitting this issue all the time now and it's making Pachyderm unusable. Pipelines get stuck in running or merging state with logs like below:

```
worker: watch closed or error running the worker process: acquire/process/merge datums for job 16310728356e403e8bcf73a0245f2811 exited with err: error claiming/processing chunk: tagGetter: tag 5d54771990e897649557c2bf9ecefcbfd4774862861edcc1135a4923943f9f9a15a0 not found; retrying in 954.512469ms
processing job 16310728356e403e8bcf73a0245f2811
blocking on parent commit "a2cb40d4898f432d87e65ee17bb792c5" before writing to output commit "fc36389b5b154b47ad07e52ed3afb3c8"
```

This just keeps repeating indefinitely and any further processing is halted. Is there anything I can do to help debug this further and find a solution?

**Comment by JoeyZwicker** [comment]
@itssimon Really sorry you're still hitting this! We're actively looking into it and will get back to you asap with some suggestions.

**Comment by brycemcanally** [comment]
Hey @itssimon, I have a few questions that I need answered before I can figure out what is going on here, you seemed to have gotten Pachyderm into a weird state.

1. Would you be able to give me some more details about the sequence of operations that seem to have led to these issues. I know you mentioned reprocess but are there any other operations such as deleting a commit, garbage collection, or enabling/disabling of pipeline features that you think might be worth mentioning?
2. Did the error messages with the substring `error claiming/processing chunk: tagGetter:` occur during the job(s) that were triggered when reprocessing?
3. Do the logs seem to be hanging at the `blocking on parent commit "a2cb40d4898f432d87e65ee17bb792c5" before writing to output commit "fc36389b5b154b47ad07e52ed3afb3c8"` error message?

**Comment by itssimon** [comment]
@brycemcanally see my answers below:

1. The sequence of operations has varied, but I think it almost certainly happens after updating a pipeline with --reprocess. However it has also happened with a job that was triggered through a normal commit to an input repo. Maybe worth mentioning is that I'm using pachypy to interact with Pachyderm (i.e. commits and create/update pipelines). pachypy uses the most up-to-date protobufs from python_pachyderm. Also worth mentioning that this only started happening with 1.9.0.
2. Yes, and it always happens at the end of that job after (almost?) all datums have been processed.
3. Yes, logs hang for about 20 seconds after that message.

Btw. I'm also seeing a lot of messages like below, all within a second:

```
error downloading chunk 120 from worker at 10.95.8.32 (rpc error: code = Unknown desc = key 120 not found in cache), falling back on object storage
...
error downloading chunk 660 from worker at 10.95.8.32 (rpc error: code = Unknown desc = key 660 not found in cache), falling back on object storage
error downloading chunk 780 from worker at 10.95.8.121 (rpc error: code = Unknown desc = key 780 not found in cache), falling back on object storage
worker: watch closed or error running the worker process: acquire/process/merge datums for job a1533b155dd7460cb9345d6fbd9b22d7 exited with err: rpc error: code = Unknown desc = tagGetter: tag aa11e8b5ca4719b6ac3f226023afc07271b0e6986ee3d1d7ca0272fbfcba1e4796ae not found; retrying in 16.59668089s
```

Another weird thing is that Pachyderm reports to have skipped a lot of datums in a job triggered by updating a pipeline with --reprocess. My understanding of this feature is that it should reprocess all datums.

**Comment by danielefranceschi** [comment]
Same error here (K8S 1.15.0+pach 1.9.0+zenko cloudserver on-premise)
Worker instance 1 is stuck at
```
2019-07-24T14:53:19.637134599Z beginning to run user code
2019-07-24T14:55:11.514863524Z Launching pipeline master process
2019-07-24T14:55:11.584679193Z found existing job 07933bfdf3ee4ca499a283af6ba29a7f for output commit cbc9a19b505f4a93bb68174c7c25a6de
2019-07-24T14:55:11.584873627Z waiting on job 07933bfdf3ee4ca499a283af6ba29a7f (pipeline version: 1, state: JOB_RUNNING)
```
Worker instance 2 is stuck like @itssimon at
```
2019-07-24T14:55:00.838569542Z finished downloading data after 47.215093ms
2019-07-24T14:55:00.838944652Z beginning to run user code
2019-07-24T14:55:46.695536854Z error downloading chunk 82 from worker at 10.40.0.6 (rpc error: code = Unknown desc = key 82 not found in cache), falling back on object storage
2019-07-24T14:56:17.705988147Z error downloading chunk 123 from worker at 10.46.0.3 (context deadline exceeded), falling back on object storage
......
```
Worker instance 3 is stuck at
```
2019-07-24T14:55:27.921694659Z" starting to download data
2019-07-24T14:55:27.967994673Z finished downloading data after 46.05622ms
2019-07-24T14:55:27.968295210Z beginning to run user code
```
After job stopped at 656/1024 datums, I've restarted the `pipeline-xxx` pods which were all stuck at `beginning to run user code`.  
Inspection of (restarted) worker containers leads to:
- running but silent python payload (my cmd)
- `/pfs/in` with one datum (a csv chunk in my case)

**Comment by brycemcanally** [comment]
Hey @itssimon, would you be able to get me a full set of logs (worker pod logs, both the user and storage containers) as well as list job and list commit output for the pipelines with issues. You can send them through Slack if you prefer. Also, reprocess should not skip any datums, but I think there might be an explanation for why we seem to be skipping datums after you enable reprocess which involves us trying to finish processing a prior job that got into a bad state.

@danielefranceschi I'm not sure if there is an actual issue based on the logs you gave. The beginning to run user code indicates the start of your code. The `error downloading chunk` is not necessarily an issue, because this can show up based on the order in which the workers quit the work for the job when it was killed (this message should probably be relabeled from error to warning or info). In @itssimon case there is an actual issue falling back on object storage when collecting chunks.

**Comment by JoeyZwicker** [comment]
@itssimon just wanted to check in here to see if you have any additional logs or info you can provide?

**Comment by itssimon** [comment]
Sorry for the delay. I've sent you an extract of the logs on Slack!

**Comment by jdoliner** [comment]
I believe this has been fixed by #3970, feel free to reopen if that's not the case.
